packages/language/src/validation/validator.ts

```diff
@@ -14,23 +14,88 @@ import {
   IRecognitionException,
   MismatchedTokenException,
 } from "chevrotain";
-import { SourceFile } from "../workspace/source-file";
+import {
+  Diagnostic,
+  DiagnosticInfo,
+  Range,
+  Severity,
+} from "../language-server/types";
 import { ReferencesCache } from "../linking/resolver";
-import { Diagnostic, Range, Severity } from "../language-server/types";
 import { isValidToken } from "../linking/tokens";
+import { PliProgram, SyntaxKind, SyntaxNode } from "../syntax-tree/ast";
+import { forEachNode } from "../syntax-tree/ast-iterator";
+import {
+  PliValidationChecks,
+  PliValidationFunction,
+  registerValidationChecks,
+} from "./pli-validator";
+
+/**
+ * A function that accepts a diagnostic for PL/I validation
+ */
+export type PliValidationAcceptor = (
+  severity: Severity,
+  message: string,
+  info: DiagnosticInfo,
+) => void;
+
+/**
+ * Generates validation diagnostics (semantic checks) from the given AST node.
+ */
+export function generateValidationDiagnostics(root: PliProgram): Diagnostic[] {
+  // TODO @montymxb Mar. 27th, 2025: Checks are generated on each invocation, not ideal, needs a rework still
```

Comment:
FYI: The `SourceFileDiagnostics` interface contains a `validation` list that should be used to store these validation diagnostics :)